src/opnsense/www/js/widgets/IpsecLeases.js

```diff
@@ -80,12 +87,54 @@ export default class IpsecLeases extends BaseTableWidget {
         $('#ipsecLeaseTable').empty().append($error);
     }
 
+    // Checks if the lease data has changed to prevent unnecessary updates
+    dataHasChanged(newLeases) {
```

Comment:
@swhite2 How about something like this? It's still updating the whole table in the UI, yet only when there are actual changes in the collected data.

I left the console log statements in so it can be verified, I can clean them up if this would be a good approach and add the same to the other ipsec widget too.

Here are the logs to verify its working:

<details>

```
IpsecLeases.js:59 Widget tick initiated.
IpsecLeases.js:62 IPsec status: {enabled: true}
IpsecLeases.js:70 IPsec leases data: {pools: {…}, leases: Array(7)}
IpsecLeases.js:126 Processing leases: (7) [{…}, {…}, {…}, {…}, {…}, {…}, {…}]
IpsecLeases.js:92 Checking if data has changed...
IpsecLeases.js:99 New leases map: {test@vpn1.test.org-172.16.203.1: {…}, test@vpn1.test.org-2003:a:1704:6363::1: {…}, lolwut@example.org-172.16.203.2: {…}, guest@example.com-192.168.0.101: {…}, admin@company.net-10.0.0.5: {…}, …}
IpsecLeases.js:100 Current leases: {}
IpsecLeases.js:112 Change detected for test@vpn1.test.org-172.16.203.1: New lease - {"pool":"pool-rw-ipv4","address":"172.16.203.1","online":false,"user":"test@vpn1.test.org"}, Current lease - undefined
IpsecLeases.js:118 Leases changed: true
IpsecLeases.js:132 Data change detected, updating UI...
IpsecLeases.js:189 Updating the HTML table with sorted rows.
IpsecLeases.js:191 HTML table updated.
IpsecLeases.js:194 Activating tooltips for new dynamic elements.
IpsecLeases.js:196 Tooltips activated.
IpsecLeases.js:59 Widget tick initiated.
IpsecLeases.js:62 IPsec status: {enabled: true}
IpsecLeases.js:70 IPsec leases data: {pools: {…}, leases: Array(7)}
IpsecLeases.js:126 Processing leases: (7) [{…}, {…}, {…}, {…}, {…}, {…}, {…}]
IpsecLeases.js:92 Checking if data has changed...
IpsecLeases.js:99 New leases map: {test@vpn1.test.org-172.16.203.1: {…}, test@vpn1.test.org-2003:a:1704:6363::1: {…}, lolwut@example.org-172.16.203.2: {…}, guest@example.com-192.168.0.101: {…}, admin@company.net-10.0.0.5: {…}, …}
IpsecLeases.js:100 Current leases: {test@vpn1.test.org-172.16.203.1: {…}, test@vpn1.test.org-2003:a:1704:6363::1: {…}, lolwut@example.org-172.16.203.2: {…}, guest@example.com-192.168.0.101: {…}, admin@company.net-10.0.0.5: {…}, …}
IpsecLeases.js:118 Leases changed: false
IpsecLeases.js:128 No data change detected.
IpsecLeases.js:59 Widget tick initiated.
IpsecLeases.js:62 IPsec status: {enabled: true}
IpsecLeases.js:70 IPsec leases data: {pools: {…}, leases: Array(7)}
IpsecLeases.js:126 Processing leases: (7) [{…}, {…}, {…}, {…}, {…}, {…}, {…}]
IpsecLeases.js:92 Checking if data has changed...
IpsecLeases.js:99 New leases map: {test@vpn1.test.org-172.16.203.1: {…}, test@vpn1.test.org-2003:a:1704:6363::1: {…}, lolwut@example.org-172.16.203.2: {…}, guest@example.com-192.168.0.101: {…}, admin@company.net-10.0.0.5: {…}, …}
IpsecLeases.js:100 Current leases: {test@vpn1.test.org-172.16.203.1: {…}, test@vpn1.test.org-2003:a:1704:6363::1: {…}, lolwut@example.org-172.16.203.2: {…}, guest@example.com-192.168.0.101: {…}, admin@company.net-10.0.0.5: {…}, …}
IpsecLeases.js:118 Leases changed: false
IpsecLeases.js:128 No data change detected.
IpsecLeases.js:59 Widget tick initiated.
IpsecLeases.js:62 IPsec status: {enabled: true}
IpsecLeases.js:70 IPsec leases data: {pools: {…}, leases: Array(7)}
IpsecLeases.js:126 Processing leases: (7) [{…}, {…}, {…}, {…}, {…}, {…}, {…}]
IpsecLeases.js:92 Checking if data has changed...
IpsecLeases.js:99 New leases map: {test@vpn1.test.org-172.16.203.1: {…}, test@vpn1.test.org-2003:a:1704:6363::1: {…}, lolwut@example.org-172.16.203.2: {…}, guest@example.com-192.168.0.101: {…}, admin@company.net-10.0.0.5: {…}, …}
IpsecLeases.js:100 Current leases: {test@vpn1.test.org-172.16.203.1: {…}, test@vpn1.test.org-2003:a:1704:6363::1: {…}, lolwut@example.org-172.16.203.2: {…}, guest@example.com-192.168.0.101: {…}, admin@company.net-10.0.0.5: {…}, …}
IpsecLeases.js:112 Change detected for admin@company.net-10.0.0.5: New lease - {"pool":"pool-rw-ipv4","address":"10.0.0.5","online":true,"user":"admin@company.net"}, Current lease - {"pool":"pool-rw-ipv4","address":"10.0.0.5","online":false,"user":"admin@company.net"}
IpsecLeases.js:118 Leases changed: true
IpsecLeases.js:132 Data change detected, updating UI...
IpsecLeases.js:189 Updating the HTML table with sorted rows.
IpsecLeases.js:191 HTML table updated.
IpsecLeases.js:194 Activating tooltips for new dynamic elements.
IpsecLeases.js:196 Tooltips activated.
```

</details>

Reply:
Here are some performance numbers with 60 fake leases as data:

1. Data has changed on Tick
```
0.5 ms 0.6 %	66.4 ms 75.3 % onWidgetTick
```

2. Data has not changed on Tick
```
0.5 ms 2.1 %	1.8 ms 7.6% onWidgetTick
```

So I assume, that when the IPsec Leases do not change on every Tick (which is improbable unless there is a very high amount of leases and a lot of users logging on and off every Tick) this new method safes a lot of processing time on the client.